src/applications/model/gaming-streaming-server.h

```diff
@@ -0,0 +1,109 @@
+/* -*- Mode:C++; c-file-style:"gnu"; indent-tabs-mode:nil; -*- */
+/*
+ * Copyright (c) 2020, University of Padova, Department of Information
+ * Engineering, SIGNET Lab.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation;
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ *
+ * Authors: Salman Mohebi <s.mohebi22@gmail.com>
+ *
+ */
+
+#ifndef GAMING_STREAMING_SERVER_H
+#define GAMING_STREAMING_SERVER_H
+
+#include "ns3/application.h"
+#include "ns3/event-id.h"
+#include "ns3/ptr.h"
+#include "ns3/ipv4-address.h"
+#include "ns3/random-variable-stream.h"
+#include "ns3/gaming-traffic-model.h"
+#include "ns3/callback.h"
+#include <functional>
+
+namespace ns3 {
+
+class Socket;
+class Packet;
+
+class GamingStreamingServer : public Application
+{
+public:
+  /**
+   * \brief Get the type ID.
+   * \return the object TypeId
+   */
+  static TypeId GetTypeId (void);
+
+  GamingStreamingServer ();
+
+  GamingStreamingServer (Ipv4Address ip, uint16_t port);
+  ~GamingStreamingServer ();
+
+  /**
+   * \brief set the remote address and port
+   * \param ip remote IP address
+   * \param port remote port
+   */
+  void SetRemote (Address ip, uint16_t port);
+
+  /**
+   * \brief set the remote address
+   * \param addr remote address
+   */
+  void SetRemote (Address addr);
+
+  /**
+   * \berief return the number of total sent packets
+   */
+  uint32_t GetTotSentPackets (void);
+
+  /**
+   * Erase the statistics of sent packets
+   */
+  void EraseStatistics (void);
+
+protected:
+  /**
+   * \brief Generate and send packets based on the different distribution
+   * \param pktDist random variable which defines the size of generated packets
+   * \param iatDist random variable which defines inter-arrival times between the packets
+   */
+  void SendPacket (Ptr<RandomVariableStream> pktDist, Ptr<RandomVariableStream> iatDist);
+
+private:
+  virtual void StartApplication (void);
+
+  virtual void StopApplication (void);
+  /**
+   * \brief Send a packet
+   */
+  virtual void Send (void);
+
+  uint32_t m_sentSeq;  //!< Sequence number for packets
+  uint32_t m_totSentPackets; //!< Counter for sent packets
+  Ptr<Socket> m_socket; //!< Socket
+  Address m_peerAddress; //!< Remote peer address
+  uint16_t m_peerPort; //!< Remote peer port
+  EventId m_sendEvent; //!< Event to send the next packet
```

Comment:
what should be in common with the children classes should be protected instead of private